### PR TITLE
Testbed improvement + WorkspaceOps fetch on missing data 

### PR DESCRIPTION
**Pull request by touilleMan**

### CODE REVIEW

File: libparsec/crates/testbed/src/template/crc_hash.rs [hunk 1]
```diff
@@ -73,20 +73,24 @@ impl<T: CrcHash> CrcHash for Vec<T> {
...
 }
 
-impl<A: CrcHash, B: CrcHash> CrcHash for HashMap<A, B> {
+impl<A: CrcHash + std::cmp::Ord, B: CrcHash> CrcHash for HashMap<A, B> {
     fn crc_hash(&self, hasher: &mut crc32fast::Hasher) {
         hasher.update(b"HashMap");
-        for (k, v) in self.iter() {
+        let mut items: Vec<_> = self.iter().collect();
+        items.sort_unstable_by(|(a, _), (b, _)| a.partial_cmp(b).unwrap());
+        for (k, v) in items {
```
Comment (FirelightFlagboy) [inline libparsec/crates/testbed/src/template/crc_hash.rs:81]: Could you use [Itertool::sorted_unstable_by](https://docs.rs/itertools/0.11.0/itertools/trait.Itertools.html#method.sorted_unstable_by) ?

File: libparsec/crates/testbed/src/template/crc_hash.rs [hunk 2]
```diff
@@ -73,20 +73,24 @@ impl<T: CrcHash> CrcHash for Vec<T> {
...
 }
 
-impl<T: CrcHash> CrcHash for HashSet<T> {
+impl<T: CrcHash + std::cmp::Ord> CrcHash for HashSet<T> {
     fn crc_hash(&self, hasher: &mut crc32fast::Hasher) {
         hasher.update(b"HashSet");
-        for v in self.iter() {
+        let mut items: Vec<_> = self.iter().collect();
+        items.sort_unstable();
+        for v in items {
```
Comment (FirelightFlagboy) [inline libparsec/crates/testbed/src/template/crc_hash.rs:93]: Could you use [Itertool::sorted_unstable](https://docs.rs/itertools/0.11.0/itertools/trait.Itertools.html#method.sorted_unstable) ?

File: libparsec/crates/testbed/src/template/build.rs [hunk 3]
```diff
@@ -175,31 +175,82 @@ impl TestbedTemplateBuilderCounters {
...
+        &mut self,
+        key: &'static str,
+        obj: &(impl std::any::Any + Clone + Send + Sync),
+    ) {
+        let boxed = Box::new(obj.to_owned());
+        // It's no big deal to leak the data here: the template is kept until the end
+        // of the program anyway (and the amount of leak is negligeable).
+        // On the other hand it allows to provide the stuff as `&'static Foo` which
+        // is convenient.
+        self.stuff.push((key, Box::leak(boxed)));
```
[the remark this answers is not in the document]
Comment (touilleMan) [reply libparsec/crates/testbed/src/template/build.rs:220]: No risk of memory increase given this is only for testbed template (that got initialized once), not testbed env (that got initialized for each test)

File: libparsec/crates/testbed/src/template/crc_hash.rs (hunk 1 above)
Comment (touilleMan) [reply libparsec/crates/testbed/src/template/crc_hash.rs:81]: This is pretty simple code, so I don't see the need to use yet another dependency

(and the [itertool version just do this under the hood anyway](https://docs.rs/itertools/0.11.0/src/itertools/lib.rs.html#2709-2711))

File: libparsec/crates/testbed/src/template/crc_hash.rs (hunk 2 above)
Comment (touilleMan) [reply libparsec/crates/testbed/src/template/crc_hash.rs:93]: see https://github.com/Scille/parsec-cloud/pull/5241#discussion_r1329896891

File: libparsec/crates/testbed/src/template/crc_hash.rs (hunk 1 above)
Comment (FirelightFlagboy) [reply libparsec/crates/testbed/src/template/crc_hash.rs:81]: We already have itertool has a dependency

**Pull request merged**